File: Code Wars Reverse and Combine Words.py
"""
Input: String containing different "words" separated by spaces.

1. More than one word? Reverse each word and combine first with seconds,
third with fourth and so on.
An odd number of words means last one stays alone but has to be reversed too.

2. Start it again until there's only one word without spaces.

3. Return your result.

Ex:
"abc def" => "cbafed"
"""


def reverse_word(word):
    lst_word = [char for char in word]
    lst_word = reversed(lst_word)
    return ''.join(lst_word)


def reverse_word_manually(word):
    reversed_word = ""

    for index in range((len(word) - 1), -1, -1):
        reversed_word += word[index]

    return reversed_word


def combine_words(lst_words):
    lst_new_words = []

    for index in range(0, len(lst_words), 2):
        try:
            lst_new_words.append(lst_words[index] + lst_words[index + 1])
        except Exception as e:
            lst_new_words.append(lst_words[index])
    return lst_new_words


def reverse_string(string):
    lst_string = string.split(' ')
    print("\n*Function Start\nSplit Input:\n", lst_string)

    while(len(lst_string) != 1):

        for index in range(len(lst_string)):
            lst_string[index] = reverse_word_manually(lst_string[index])

        lst_string = combine_words(lst_string)
        print("Input at length", len(lst_string), ':', lst_string)

    return ''.join(lst_string)


input_2 = "abc def"
input_3 = "abc def hij"
input_6 = "abc def hij abc def hij"

for string in [input_2, input_3, input_6]:
    print("""\nInput: "{}":\n"{}" """.format(string, reverse_string(string)))

# print("""\nInput: "{}":""".format(reverse_string())
# print("""\nInput: "{}":""".format(reverse_string(input_3)))
# print("\nInput:", input_6, ':', reverse_string(input_6))

File: Code Wars Reverse and Combine Words.jl
"""
Input: String containing different "words" separated by spaces.

1. More than one word? Reverse each word and combine first with second,
third with fourth and so on.
An odd number of words means last one stays alone but has to be reversed too.

2. Start it again until there's only one word without spaces.

3. Return your result.

Ex:
"abc def" => "cbafed"
"""


function reverse_strings(string::String)
    string_array = split(string)
    iterations_needed = ceil(length(string_array)/2)
    println("Iterations needed:", iterations_needed)    
    iterations_done = 0
    println("Iterations done:", iterations_done)

    #while (length(string_array) != 1)
    while (iterations_done != iterations_needed)
        new_array = []
        array_length = length(string_array)

        for index in 1:array_length
            string_array[index] = reverse(string_array[index])            
        end

        for index in range(1, array_length, step = 2)
            if !((index + 1) > length(string_array))
                joined_words = join([string_array[index], string_array[index + 1]])
                push!(new_array, joined_words)
            else
                push!(new_array, string_array[index])
            end
        end

        #println("String array: ", string_array)
        string_array = new_array
        iterations_done += 1

    end

    return join(string_array)

end


function join_pairs!(string_array::Array)
    new_array = []

    for index in range(1, length(string_array), step = 2)

        if !((index + 1) > length(string_array))
            joined_pair = join([string_array[index], string_array[index + 1]])
            push!(new_array, joined_pair)
        else
            push!(new_array, string_array[index])
        end

    end

    return new_array

end


function reverse_strings_b(string::String)
    string_array = split(string)        

    while (length(string_array) != 1)            
        array_length = length(string_array)

        for index in 1:array_length
            string_array[index] = reverse(string_array[index])            
        end

        #for word in string_array
         #   word = reverse(word)
        #end

        string_array = join_pairs!(string_array)
        #println("String array: ", string_array)        

    end

    return join(string_array)

end

input_2 = "abc def"
input_3 = "abc def hij"
input_6 = "abc def hij abc def hij"

println("Input `$input_2`: ", reverse_strings(input_2))
println("Input `$input_3`: ", reverse_strings(input_3))
println("Input `$input_6`: ", reverse_strings(input_6))

println("Called function with separate joining function with `$input_2`:\n", reverse_strings_b(input_2))
println("Called function with separate joining function with `$input_3`:\n", reverse_strings_b(input_3))
println("Called function with separate joining function with `$input_6`:\n", reverse_strings_b(input_6))
